libs/sqlserver/langchain_sqlserver/vectorstores.py

```diff
@@ -225,31 +229,86 @@ def __init__(
         self._embedding_store = self._get_embedding_store(self.table_name, self.schema)
         self._create_table_if_not_exists()
 
-    def _can_connect_with_entra_id(self) -> bool:
-        """Determine if Entra ID authentication can be used.
+    def _get_connection_url(self, conn_string: str) -> str:
+        if conn_string is None or len(conn_string) == 0:
+            logging.error("Connection string value is None or empty.")
+            raise ValueError("Connection string value cannot be None.")
+
+        args = conn_string.split(";")
+        arg_dict = {}
+        for arg in args:
+            if "=" in arg:
+                # Split into key value pairs by the first positioned `=` found.
+                # Key-Value pairs are inserted into the dictionary.
+                #
+                key, value = arg.split("=", 1)
+                arg_dict[key.lower().strip()] = value.strip()
 
-        Check the components of the connection string to determine
-        if connection via Entra ID authentication is possible or not.
+        try:
+            # This will throw a key error if server or database keyword
+            # is not present in arg_dict from the connection string.
+            #
+            database = arg_dict.pop("database")
 
-        The connection string is of expected to be of the form:
-            "mssql+pyodbc://username:password@servername/dbname?other_params"
-        which gets parsed into -> <scheme>://<netloc>/<path>?<query>
-        """
-        parsed_url = urlparse(self.connection_string)
+            # If `server` is present in the dictionary, we split by
+            # `,` to obtain host and port details.
+            #
+            server = arg_dict.pop("server").split(",", 1)
+            server_host = server[0]
```

Comment:
Are we not checking if server_host is invalid data similar to server_port below?

Reply:
The purpose of this function is to create a string that's in a format that sqlalchemy will be able to parse. Every validation we do here is done so that whatever process we do on passed in data doesn't fail. For actual value validation, sqlalchemy will handle that. If an invalid host is passed in, connection will fail.

For server port, the validation we have is done to know if we can cast the provided value to an integer. Whether the eventual value is valid or not we don't know yet - handled by sqlalchemy, the connection library.